### ISSUE TITLE
Idea:  Let loading state be handled by `render`?

### ISSUE THREAD

**Issue by ianstormtaylor** [facebook/relay#821]
I was a bit confused as to why the `renderLoading` function needed to be separate from the `render` method of the `Component` passed to `RootContainer`. It seemed like something that would be nicer to have handled as a `props.loading = true` case (or even `props.relay.loading` if namespacing is an issue), and have all of the rendering logic for that top-level component be handled in a single place, instead of separately having rendering logic in what is affectively the “router” of an application. 

Same goes for `renderFailure` that could be handled via `props.relay.error` or similar.

Similarly I’m not sure why the `renderFetched` needs to exist, although it definitely seems like there could be valid use cases. It might be simpler to just have users handle that logic in userspace, by wrapping whatever component they wanted with a parent one that proxies.

I feel like I might be missing something here, since I haven't needed to deal with any complex cases here. But it just seemed like an odd thing when getting familiarized.

**Comment by josephsavona** [comment]
>  was a bit confused as to why the renderLoading function needed to be separate from the render method of the Component passed to RootContainer

Relay was built to simplify the process of loading data for a view hierarchy. To that end, we designed it so that individual containers/components do not have to be concerned with their initial loading state - Relay will ensure that they are not rendered until their data is available. Without this guarantee, *every* component would have to handle the theoretical possibility of being in a "loading" state - which is extra boilerplate that should only apply at the top of the view. We've found that the current approach simplifies the process of writing larger complex apps - there's a bit more background about this in [Thinking in Relay](http://facebook.github.io/relay/docs/thinking-in-relay.html#fetching-data-for-a-view). 

The existing callbacks - `renderFetched, `renderLoading`, etc - are certainly confusing. In an upcoming release we will deprecate `<RelayRootContainer>` and offer access to `<RelayRenderer>`, which consolidates all of these props into a [single `render` callback](https://github.com/facebook/relay/blob/master/src/container/RelayRenderer.js#L80).

Again, we really appreciate your feedback here. Let's discuss this a bit more in your issue about removing routes as a concept (since the handling of loading states would have to move somewhere in that case).
